generator/mwm_diff/diff.cpp

```diff
@@ -2,53 +2,121 @@
+#include "coding/write_to_sink.hpp"
+#include "coding/writer.hpp"
+#include "coding/zlib.hpp"
 
 #include "base/logging.hpp"
 
+#include <cstdint>
+#include <iterator>
+#include <vector>
+
+#include "3party/bsdiff-courgette/bsdiff/bsdiff.h"
+
 using namespace std;
 
 namespace
 {
-// We could use my::CopyFileX but it fails if the file at |srcFile| is empty.
-bool CopyFile(string const & srcFile, string const & dstFile)
+// Format Version 0: bsdiff+gzip.
+uint32_t const kLatestVersion = 0;
+}  // namespace
+
+namespace generator
+{
+namespace mwm_diff
+{
+bool MakeDiff(string const & oldMwmPath, string const & newMwmPath, string const & diffPath)
 {
   try
   {
-    FileReader reader(srcFile, true /* withExceptions */);
-    ReaderSource<FileReader> src(reader);
-    FileWriter writer(dstFile);
+    FileReader oldReader(oldMwmPath);
+    FileReader newReader(newMwmPath);
+    FileWriter diffFileWriter(diffPath);
+
+    vector<uint8_t> diffBuf;
+    MemWriter<vector<uint8_t>> diffMemWriter(diffBuf);
+
+    auto const status = bsdiff::CreateBinaryPatch(oldReader, newReader, diffMemWriter);
 
-    rw::ReadAndWrite(src, writer);
+    if (status != bsdiff::BSDiffStatus::OK)
+    {
+      LOG(LERROR, ("Could not create patch with bsdiff:", status));
+      return false;
+    }
+
+    using Deflate = coding::ZLib::Deflate;
+    Deflate deflate(Deflate::Format::ZLib, Deflate::Level::BestCompression);
+
+    vector<uint8_t> deflatedDiffBuf;
+    deflate(diffBuf.data(), diffBuf.size(), back_inserter(deflatedDiffBuf));
+
+    uint32_t const header = kLatestVersion;
+    WriteToSink(diffFileWriter, header);
+    diffFileWriter.Write(deflatedDiffBuf.data(), deflatedDiffBuf.size());
   }
   catch (Reader::Exception const & e)
   {
-    LOG(LERROR, ("Could not read from", srcFile, ":", e.Msg()));
+    LOG(LERROR, ("Could not open file when creating a patch:", e.Msg()));
     return false;
   }
   catch (Writer::Exception const & e)
   {
-    LOG(LERROR, ("Could not write to", dstFile, ":", e.Msg()));
+    LOG(LERROR, ("Could not open file when creating a patch:", e.Msg()));
     return false;
   }
 
   return true;
 }
-}  // namespace
 
-namespace generator
+bool ApplyDiff(string const & oldMwmPath, string const & newMwmPath, string const & diffPath)
 {
-namespace mwm_diff
-{
-bool MakeDiff(string const & /* oldMwmPath */, string const & newMwmPath, string const & diffPath)
-{
-  return CopyFile(newMwmPath, diffPath);
-}
+  try
+  {
+    FileReader oldReader(oldMwmPath);
+    FileWriter newWriter(newMwmPath);
+    FileReader diffFileReader(diffPath);
 
-bool ApplyDiff(string const & /* oldMwmPath */, string const & newMwmPath, string const & diffPath)
-{
-  return CopyFile(diffPath, newMwmPath);
+    ReaderSource<FileReader> diffFileSource(diffFileReader);
+    auto const header = ReadPrimitiveFromSource<uint32_t>(diffFileSource);
+    if (header != kLatestVersion)
+    {
+      LOG(LERROR, ("Unknown version format of mwm diff:", header));
```

Comment:
We may try to apply not only latest-version diffs. In general, your code should always generate diff in the latest format, but must be able to apply diffs of all formats.